lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -69,6 +74,71 @@ LogicalResult CreateSiFiveMetadataPass::emitMemoryMetadata() {
       !dutMod ||
       instancePathCache.instanceGraph.getTopLevelNode()->getModule() == dutMod;
 
+  auto *context = circuitOp.getContext();
+  auto builderOM = OpBuilder::atBlockEnd(circuitOp.getBodyBlock());
+
+  using NameTypePair = std::pair<StringAttr, mlir::Type>;
+  // Add all the properties of a memory as fields of the class.
+  // The types must match exactly with the FMemModuleOp attribute type.
+  SmallVector<NameTypePair> classFields;
+  classFields.push_back({StringAttr::get(context, "name"),
+                         om::StringOMType::get(builderOM.getContext())});
+  classFields.push_back(
+      {StringAttr::get(context, "depth"),
+       mlir::IntegerType::get(context, 64, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "width"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "maskBits"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "readPorts"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "writePorts"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "readwritePorts"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "writeLatency"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+  classFields.push_back(
+      {StringAttr::get(context, "readLatency"),
+       mlir::IntegerType::get(context, 32, IntegerType::Unsigned)});
+
+  auto unknownLoc = builderOM.getUnknownLoc();
+  // Get only the attribute names for parameter names.
+  auto paramNames = builderOM.getArrayAttr(llvm::to_vector(llvm::map_range(
+      classFields, [&](NameTypePair e) -> Attribute { return e.first; })));
+
+  // Memory metadata class.
+  auto memMetadataClass = builderOM.create<circt::om::ClassOp>(
+      builderOM.getUnknownLoc(),
+      StringAttr::get(builderOM.getContext(), "MemoryConf"), paramNames);
+
+  // Now add all the properties of the memory as a ClassFieldOp.
+  Block *body = new Block();
+  memMetadataClass.getRegion().push_back(body);
+  builderOM.setInsertionPointToEnd(body);
+  for (auto field : classFields)
+    builderOM.create<om::ClassFieldOp>(
+        unknownLoc, field.first, body->addArgument(field.second, unknownLoc));
```

Comment:
nit: Could you separate this part into builder function of ClassOp?

Reply:
In general we need nice builders for the OM ops. So far they have just been using the auto-generated ones from ODS. I think after looking at this PR, and other uses of the ops from C++ (e.g. unit tests), we know enough to define some helpful builders.